### PR TITLE
Do not fail worker when insufficient memory happened for starting a d…

### LINKED ISSUE #1811: Do not fail worker when there is no sufficient memory to start a docker container
This issue has happened several times in production. When there is not sufficient memory to start a docker container, the worker will throw an exception and go offline. This ticket is to change the behavior to keep the worker alive even if there is an exception thrown for a particular bundle so that the worker can still be used to run other jobs. 
An example of the error message from one production worker:
```
vm-clws-prod-gpuworker-5:~% docker logs codalab_worker_1 --tail 10
  File "/opt/codalab/worker/local_run/local_run_manager.py", line 190, in process_runs
    self._runs[bundle_uuid] = self._run_state_manager.transition(run_state)
  File "/opt/codalab/worker/fsm.py", line 35, in transition
    return self._transition_functions[state.stage](state)
  File "/opt/codalab/worker/local_run/local_run_state.py", line 292, in _transition_from_PREPARING
    runtime=self.docker_runtime,
  File "/opt/codalab/worker/docker_utils.py", line 33, in wrapper
    raise DockerException(message + ': ' + str(e))
codalab.worker.docker_utils.DockerException: Unable to start Docker container: Bundle 0xf97322adbb404e75801275d0e6cdf30b: minimum memory to start pytorch/pytorch:latest must be 4m (4194304 bytes), current available memory is 20 bytes
```

### PR DESCRIPTION
**Pull request by candicegjing**
Fixed #1811 
The final result would be look like in the following picture
Screen Shot 2019-12-03 at 1 43 29 PM

### CODE REVIEW

**Comment by yipenghe** [comment]
We should get this in soon

**Review by percyliang — COMMENTED** [review]
One concern I have is if that a worker is broken, then this change will end up failing a bunch of runs, which is worse than just essentially bringing the worker offline while we figure out what's wrong.

**Comment by yipenghe** [comment]
> One concern I have is if that a worker is broken, then this change will end up failing a bunch of runs, which is worse than just essentially bringing the worker offline while we figure out what's wrong.

Then maybe we should just reject memory request less than, say, 10m? Not sure happens when memory is enough to run start container but not enough for running a run though

**Comment by percyliang** [comment]
Ideally, we would target particular errors (like not enough memory)...not sure if we can read that out of the Exception.  Or maybe better, we should make sure we set a minimum memory before starting the container.

**Comment by yipenghe** [comment]
> Ideally, we would target particular errors (like not enough memory)...not sure if we can read that out of the Exception. Or maybe better, we should make sure we set a minimum memory before starting the container.

We probably should either 1. replace insufficient memory request with minimum memory or 2. Reject the run immediately after reading the request

**Comment by percyliang** [comment]
We should reject the request on the backend, because it seems like this only happens when someone deliberately does `--request-memory 20`.

**Comment by yipenghe** [comment]
> We should reject the request on the backend, because it seems like this only happens when someone deliberately does `--request-memory 20`.

Might also be forgot to add the unit (b, m , g etc.), but I agree we should reject it on the backend. How much should the minimum memory be, maybe like ~100m?

**Comment by percyliang** [comment]
Sure, `100m` sounds good.

**Comment by candicegjing** [comment]
> One concern I have is if that a worker is broken, then this change will end up failing a bunch of runs, which is worse than just essentially bringing the worker offline while we figure out what's wrong.

Can you explain more on why this change would fail a bunch of runs when a worker is broken? I think this change will only affect the bundle which has insufficient memory requested.

**Comment by candicegjing** [comment]
> Ideally, we would target particular errors (like not enough memory)...not sure if we can read that out of the Exception. Or maybe better, we should make sure we set a minimum memory before starting the container.

Yes. There is a piece of code that checks if the current requested memory is sufficient enough to start a container. The exception is thrown from that logic. Also, we can target that particular errors.

**Comment by candicegjing** [comment]
Also failing the worker because of exception is thrown from one job doesn't seem to be a good idea. Right now the exception is due to insufficient memory, but that `try except` block will also catch other docker exceptions as well. Then the worker will stop accepting new jobs right away, which is not quite good.

**Comment by percyliang** [comment]
If there is an error and it's the worker's fault, then we want the worker to fail so that we don't keep on failing all jobs that are assigned to that bad worker.  If it's the job's fault, we should fail the job.

**Comment by candicegjing** [comment]
> If there is an error and it's the worker's fault, then we want the worker to fail so that we don't keep on failing all jobs that are assigned to that bad worker. If it's the job's fault, we should fail the job.

I think the confusion for this issue is that if the exception represents a broken worker or a bad job request. In this particular case, it's a bad job request (the error message might be reworded to *requested memory* instead of *available memory*), and the exception is from Docker not from the worker itself. I think rejecting the job when it requests insufficient memory is also helpful. We need to decide at which stage we want to reject it. For example, should we reject it during the time of request submission, reject it before assigning it to a worker or after assigning it to a worker? Also since there are other computing resources which will be requested from a user (CPUs, GPUs) and could potentially have invalid values like `0 CPU`, do we also consider adding a sanity check on those value and reject them early?

File: codalab/server/bundle_manager.py [hunk 1]
```diff
@@ -22,6 +22,11 @@
 SECONDS_PER_DAY = 60 * 60 * 24
 # Fail unresponsive bundles in uploading, staged and running state after this many days.
 BUNDLE_TIMEOUT_DAYS = 60
+# Impose a minimum container request memory 4mb (4096 bytes), same as docker's minimum allowed value
+# https://docs.docker.com/config/containers/resource_constraints/#limit-a-containers-access-to-memory
+# When using the REST api, it is allowed to set Memory to 0 but that means the container has unbounded
+# access to the host machine's memory, which we have decided to not allow
+MINIMUM_REQUEST_MEMORY = 4096
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Make type explicit: `MINIMUM_REQUEST_MEMORY_BYTES`

File: codalab/server/bundle_manager.py [hunk 2]
```diff
@@ -22,6 +22,11 @@
 SECONDS_PER_DAY = 60 * 60 * 24
 # Fail unresponsive bundles in uploading, staged and running state after this many days.
 BUNDLE_TIMEOUT_DAYS = 60
+# Impose a minimum container request memory 4mb (4096 bytes), same as docker's minimum allowed value
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: 4mb is not 4096 bytes

File: codalab/server/bundle_manager.py (hunk 1 above)
Comment (percyliang) [reply codalab/server/bundle_manager.py]: write `4 * 1024 * 1024`

**Review by percyliang — APPROVED** [review]

**Comment by yipenghe** [comment]
Kind of confused, how did #2514 happen after this PR?

**Pull request merged**